packages/astro/src/core/render-context.ts

```diff
@@ -282,42 +262,51 @@ export class RenderContext {
 					throw new AstroError(AstroErrorData.StaticClientAddressNotAvailable);
 				}
 			},
-			// `Astro.self` is added by the compiler
-		} as unknown as typeof astroStaticPartial &
-			typeof astroDynamicPartial &
-			Pick<AstroGlobal, 'clientAddress' | 'self'>;
+			cookies,
+			get currentLocale() { return renderContext.computeCurrentLocale(); },
+			locals,
+			params,
+			get preferredLocale() { return renderContext.computePreferredLocale(); },
+			get preferredLocaleList() { return renderContext.computePreferredLocaleList(); },
+			props,
+			redirect,
+			request,
+			response,
+			slots,
+			url,
+		};
+
+		// `Astro.self` is added by the compiler
+		return Object.setPrototypeOf(astroDynamicPartial, astroStaticPartial) as unknown as typeof astroStaticPartial & typeof astroDynamicPartial & Pick<AstroGlobal, 'self'>;
```

Comment:
Using setPrototypeOf causes code to be deoptimized. I don't know if that matters in this case, but this is a hot code-path so it very well might. Can we go back to setting `__proto__` instead?

Reply:
That required unsafe assertions. I'll take another look.